File: android/src/main/kotlin/com/hexgram/android/viewmodels/BaziViewModel.kt
package com.hexgram.android.viewmodels

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.hexgram.android.models.AIService
import com.hexgram.android.models.BaziEngine
import com.hexgram.android.models.BaziResult
import com.hexgram.android.models.LunarCalendar
import kotlinx.coroutines.launch
import java.util.Calendar

class BaziViewModel(application: Application) : AndroidViewModel(application) {

    // Input state
    var selectedYear by mutableStateOf(1990)
    var selectedMonth by mutableStateOf(1)
    var selectedDay by mutableStateOf(1)
    var selectedHourIndex by mutableStateOf(0)
    var sex by mutableStateOf("M") // "M" or "F"
    var name by mutableStateOf("")

    // 阴历模式
    var isLunar by mutableStateOf(false)
    var lunarDisplayString by mutableStateOf("")

    // Result state
    var result by mutableStateOf<BaziResult?>(null)
    var resultText by mutableStateOf("")

    // AI state
    var aiText by mutableStateOf("")
    var aiLoading by mutableStateOf(false)

    companion object {
        val CHINESE_HOURS = listOf(
            "子时 (23-01)", "丑时 (01-03)", "寅时 (03-05)", "卯时 (05-07)",
            "辰时 (07-09)", "巳时 (09-11)", "午时 (11-13)", "未时 (13-15)",
            "申时 (15-17)", "酉时 (17-19)", "戌时 (19-21)", "亥时 (21-23)"
        )
        val HOUR_VALUES = listOf(23, 1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21)
    }

    fun calculate() {
        val hourValue = HOUR_VALUES[selectedHourIndex]

        val y: Int; val m: Int; val d: Int
        if (isLunar) {
            val solar = LunarCalendar.lunarToSolar(selectedYear, selectedMonth, selectedDay)
                ?: return
            y = solar.first; m = solar.second; d = solar.third
            val lunarDate = LunarCalendar.LunarDate(selectedYear, selectedMonth, selectedDay)
            lunarDisplayString = lunarDate.displayString
        } else {
            y = selectedYear; m = selectedMonth; d = selectedDay
            lunarDisplayString = ""
        }

        val baziResult = BaziEngine.calculate(y, m, d, hourValue, sex, name)
        result = baziResult
        var text = BaziEngine.formatPlainText(baziResult)
        if (isLunar && lunarDisplayString.isNotEmpty()) {
            text = "【$lunarDisplayString】\n$text"
        }
        resultText = text
    }

    val lunarMaxDay: Int
        get() = LunarCalendar.daysInLunarMonth(selectedYear, selectedMonth)

    fun reset() {
        result = null
        resultText = ""
        aiText = ""
        selectedYear = 1990
        selectedMonth = 1
        selectedDay = 1
        selectedHourIndex = 0
        sex = "M"
        name = ""
    }

    fun requestAI() {
        if (resultText.isBlank()) return
        val endpoint = AIService.getEndpoint(getApplication())
        if (endpoint.isBlank()) {
            aiText = "未配置Worker地址。请在设置中配置后端地址。"
            return
        }
        aiLoading = true
        aiText = ""
        viewModelScope.launch {
            try {
                aiText = AIService.callWorker(
                    endpoint = endpoint,
                    type = "bazi",
                    data = resultText
                )
            } catch (e: Exception) {
                aiText = "AI解读失败：${e.message}"
            } finally {
                aiLoading = false
            }
        }
    }
}

File: ios/Hexgram/ViewModels/BaziViewModel.swift
import SwiftUI

@MainActor
class BaziViewModel: ObservableObject {
    @Published var selectedDate = Calendar.current.date(from: DateComponents(year: 1990, month: 1, day: 1))!
    @Published var selectedHour = 11
    @Published var sex = "M"
    @Published var name = ""
    @Published var result: BaziResult?
    @Published var resultText = ""

    // 阴历模式
    @Published var isLunar = false
    @Published var lunarYear = 1990
    @Published var lunarMonth = 1
    @Published var lunarDay = 1
    @Published var lunarDisplayString = ""

    let aiService = AIService()

    static let hourOptions: [(label: String, value: Int)] = [
        ("子时", 23), ("丑时", 1), ("寅时", 3), ("卯时", 5),
        ("辰时", 7), ("巳时", 9), ("午时", 11), ("未时", 13),
        ("申时", 15), ("酉时", 17), ("戌时", 19), ("亥时", 21)
    ]

    func calculate() {
        var y: Int, m: Int, d: Int

        if isLunar {
            // 阴历→公历转换
            guard let solar = LunarCalendar.lunarToSolar(year: lunarYear, month: lunarMonth, day: lunarDay) else {
                return
            }
            y = solar.year; m = solar.month; d = solar.day
            let lunarDate = LunarCalendar.LunarDate(year: lunarYear, month: lunarMonth, day: lunarDay, isLeapMonth: false)
            lunarDisplayString = lunarDate.displayString
        } else {
            let comps = Calendar.current.dateComponents([.year, .month, .day], from: selectedDate)
            guard let cy = comps.year, let cm = comps.month, let cd = comps.day else { return }
            y = cy; m = cm; d = cd
            lunarDisplayString = ""
        }

        let r = BaziEngine.calculate(
            year: y, month: m, day: d,
            hour: selectedHour, sex: sex, name: name
        )
        result = r
        var text = BaziEngine.formatPlainText(r)
        if isLunar, !lunarDisplayString.isEmpty {
            text = "【\(lunarDisplayString)】\n" + text
        }
        resultText = text
    }

    func aiRead() async {
        guard !resultText.isEmpty else { return }
        await aiService.callWorker(type: "bazi", data: resultText)
    }

    // 阴历月最大天数
    var lunarMaxDay: Int {
        LunarCalendar.daysInLunarMonth(year: lunarYear, month: lunarMonth)
    }
}
